extensions/rv32im/circuit/src/adapters/loadstore.rs

```diff
@@ -579,24 +581,43 @@ where
             opcode.local_opcode_idx(Rv32LoadStoreOpcode::CLASS_OFFSET),
         );
 
-        let read_data: [u8; RV32_REGISTER_NUM_LIMBS] = match local_opcode {
+        let rs1 = read_rv32_register_from_state(state, b.as_canonical_u32());
+        let imm_extended = c.as_canonical_u32() + g.as_canonical_u32() * 0xffff0000;
+
+        let ptr_val = rs1.wrapping_add(imm_extended);
+        let shift_amount = ptr_val & 3;
+        let ptr_val = ptr_val - shift_amount;
+
+        assert!(
+            ptr_val < (1 << self.pointer_max_bits),
+            "ptr_val: {ptr_val} = rs1_val: {} + imm_extended: {imm_extended} >= 2 ** {}",
+            rs1,
+            self.pointer_max_bits
+        );
+
+        let read_data = match local_opcode {
             LOADW | LOADB | LOADH | LOADBU | LOADHU => {
-                let rs1 = read_rv32_register_from_state(state, b.as_canonical_u32());
-                let imm_extended = c.as_canonical_u32() + g.as_canonical_u32() * 0xffff0000;
-                let ptr_val = rs1.wrapping_add(imm_extended);
-                assert!(
-                    ptr_val < (1 << self.pointer_max_bits),
-                    "ptr_val: {ptr_val} = rs1_val: {rs1} + imm_extended: {imm_extended} >= 2 ** {}",
-                    self.pointer_max_bits
-                );
-                memory_read_from_state(state, e.as_canonical_u32(), ptr_val)
+                let read_data = memory_read_from_state(state, e.as_canonical_u32(), ptr_val);
+                read_data
```

Comment:
clippy might complain
```suggestion
                memory_read_from_state(state, e.as_canonical_u32(), ptr_val)
```